File: scripts/setup_database.py
#!/usr/bin/env python3
"""
Complete database setup script.
Initializes database and populates all test cases in one command.

Usage:
    python3 scripts/setup_database.py
"""

import sqlite3
import json
import sys
import os
from pathlib import Path

def load_test_data():
    """Load test case data from JSON files"""
    script_dir = Path(__file__).parent
    data_dir = script_dir.parent / 'python-service' / 'data'

    # Load hidden test cases
    with open(data_dir / 'hidden_test_cases.json', 'r') as f:
        hidden_test_cases = json.load(f)

    # Load Python imports
    with open(data_dir / 'python_imports.json', 'r') as f:
        python_imports = json.load(f)

    return hidden_test_cases, python_imports

def get_db_path():
    """Get the database path based on OS"""
    if os.name == 'nt':  # Windows
        db_path = Path(os.environ.get('APPDATA', '')) / 'interview-prep-platform' / 'interview-prep.db'
    elif sys.platform == 'darwin':  # macOS
        db_path = Path.home() / 'Library' / 'Application Support' / 'interview-prep-platform' / 'interview-prep.db'
    else:  # Linux
        db_path = Path.home() / '.config' / 'interview-prep-platform' / 'interview-prep.db'
    return db_path

def main():
    print("=" * 70)
    print("Complete Database Setup")
    print("=" * 70)
    print()

    # Load test data from JSON files
    print("📦 Loading test data from JSON files...")
    HIDDEN_TEST_CASES, PYTHON_IMPORTS = load_test_data()
    print(f"   ✅ Loaded {len(HIDDEN_TEST_CASES)} question test cases")
    print(f"   ✅ Loaded {len(PYTHON_IMPORTS)} Python import configurations")
    print()

    db_path = get_db_path()

    if not db_path.exists():
        print(f"❌ Database not found at: {db_path}")
        print("   Please run the application first to create the database.")
        print("   Or run: npm run db:init")
        sys.exit(1)

    print(f"📂 Database: {db_path}")
    print()

    try:
        conn = sqlite3.connect(str(db_path))
        cursor = conn.cursor()

        # Update hidden test cases
        print("📝 Updating hidden test cases...")
        updated_count = 0
        for title, hidden_tests in HIDDEN_TEST_CASES.items():
            cursor.execute("""
                UPDATE leetcode_questions
                SET hidden_test_cases = ?
                WHERE question_id IN (
                    SELECT id FROM questions WHERE title = ?
                )
            """, (json.dumps(hidden_tests), title))

            if cursor.rowcount > 0:
                updated_count += 1
                print(f"   ✅ {title}")

        print(f"\n✅ Updated hidden test cases for {updated_count} questions")

        # Update Python imports
        print("\n📝 Updating Python imports...")
        imports_count = 0
        for title, imports in PYTHON_IMPORTS.items():
            cursor.execute("""
                UPDATE leetcode_questions
                SET function_signature_python = ?
                WHERE question_id IN (
                    SELECT id FROM questions WHERE title = ?
                )
            """, (imports, title))

            if cursor.rowcount > 0:
                imports_count += 1
                print(f"   ✅ {title}")

        print(f"\n✅ Updated imports for {imports_count} questions")

        # Commit changes
        conn.commit()

        # Verify setup
        print("\n🔍 Verifying setup...")
        cursor.execute("""
            SELECT COUNT(*) as total,
                   SUM(CASE WHEN json_array_length(hidden_test_cases) > 0 THEN 1 ELSE 0 END) as with_hidden
            FROM leetcode_questions
        """)

        total, with_hidden = cursor.fetchone()

        print(f"   Total questions: {total}")
        print(f"   With hidden tests: {with_hidden}")

        if with_hidden == total and total > 0:
            print("\n" + "=" * 70)
            print("✅ Database setup complete!")
            print("=" * 70)
            print("\n📊 Summary:")
            print(f"   • {total} LeetCode questions")
            print(f"   • {with_hidden} questions with hidden tests")
            print(f"   • ~{with_hidden * 3} visible test cases")
            print(f"   • ~{with_hidden * 4} hidden test cases")
            print(f"   • ~{with_hidden * 7} total test cases")
            print("\n🚀 Next steps:")
            print("   1. Set execution mode in .env (SANDBOX_MODE=local or docker)")
            print("   2. Start the app: npm run dev")
            print("   3. Try submitting a solution!")
            sys.exit(0)
        else:
            print("\n❌ Some questions missing hidden tests")
            sys.exit(1)

    except Exception as e:
        print(f"\n❌ Error: {e}")
        import traceback
        traceback.print_exc()
        sys.exit(1)
    finally:
        conn.close()

if __name__ == "__main__":
    main()

File: scripts/setupDatabase.js
#!/usr/bin/env node
/**
 * Database Setup Script
 *
 * This script:
 * 1. Initializes the database with the schema
 * 2. Seeds it with all question data (including ML design solutions)
 *
 * Usage: node scripts/setupDatabase.js
 */

const Database = require('better-sqlite3');
const fs = require('fs');
const path = require('path');
const os = require('os');

// Determine database location based on platform
function getDatabasePath() {
  const appName = 'interview-prep-platform';
  let configDir;

  if (process.platform === 'win32') {
    configDir = path.join(process.env.APPDATA || '', appName);
  } else if (process.platform === 'darwin') {
    configDir = path.join(os.homedir(), 'Library', 'Application Support', appName);
  } else {
    configDir = path.join(os.homedir(), '.config', appName);
  }

  if (!fs.existsSync(configDir)) {
    fs.mkdirSync(configDir, { recursive: true });
  }

  return path.join(configDir, 'interview-prep.db');
}

async function setupDatabase() {
  try {
    const dbPath = getDatabasePath();
    console.log('📦 Setting up database at:', dbPath);
    console.log('');

    // Check if database already exists
    const dbExists = fs.existsSync(dbPath);
    if (dbExists) {
      console.log('⚠️  Database already exists. This will overwrite existing data.');
      console.log('   Backing up existing database...');
      const backupPath = dbPath + '.backup.' + Date.now();
      fs.copyFileSync(dbPath, backupPath);
      console.log('   Backup created at:', backupPath);
      console.log('');
    }

    const db = new Database(dbPath);
    db.pragma('journal_mode = WAL');
    db.pragma('foreign_keys = ON');

    // Step 1: Initialize schema
    console.log('📋 Step 1: Initializing database schema...');
    const schemaPath = path.join(__dirname, '../database/schema.sql');
    const schema = fs.readFileSync(schemaPath, 'utf-8');
    db.exec(schema);
    console.log('   ✓ Schema initialized');
    console.log('');

    // Step 2: Import seed data
    console.log('📥 Step 2: Importing question data...');
    const seedPath = path.join(__dirname, '../database/seed_complete.sql');
    console.log('   Reading seed file:', seedPath);

    const seedData = fs.readFileSync(seedPath, 'utf-8');
    console.log('   Seed file size:', (seedData.length / 1024).toFixed(2), 'KB');
    console.log('   Executing SQL statements... (this may take a moment)');

    db.exec(seedData);
    console.log('   ✓ Seed data imported');
    console.log('');

    // Step 3: Verify data was imported correctly
    console.log('✅ Step 3: Verifying data import...');

    const questionCount = db.prepare('SELECT COUNT(*) as count FROM questions').get();
    const leetcodeCount = db.prepare('SELECT COUNT(*) as count FROM leetcode_questions').get();
    const mlDesignCount = db.prepare('SELECT COUNT(*) as count FROM ml_design_questions').get();

    console.log('   - Total questions:', questionCount.count);
    console.log('   - LeetCode questions:', leetcodeCount.count);
    console.log('   - ML Design questions:', mlDesignCount.count);

    // Check ML design questions with solutions
    const mlWithSolutions = db.prepare(`
      SELECT COUNT(*) as count
      FROM ml_design_questions
      WHERE sample_solution IS NOT NULL AND sample_solution != ''
    `).get();

    console.log('   - ML Design questions with solutions:', mlWithSolutions.count);

    // Show sample solution lengths
    const solutionLengths = db.prepare(`
      SELECT
        q.title,
        LENGTH(ml.sample_solution) as solution_length
      FROM ml_design_questions ml
      JOIN questions q ON ml.question_id = q.id
      WHERE ml.sample_solution IS NOT NULL AND ml.sample_solution != ''
      ORDER BY ml.question_id
      LIMIT 5
    `).all();

    if (solutionLengths.length > 0) {
      console.log('');
      console.log('   Sample solution lengths (first 5):');
      solutionLengths.forEach(s => {
        const lengthKB = (s.solution_length / 1024).toFixed(2);
        console.log(`   - ${s.title}: ${s.solution_length} chars (${lengthKB} KB)`);
      });
    }

    db.close();

    console.log('');
    console.log('🎉 Database setup complete!');
    console.log('');
    console.log('You can now run the application with: npm run dev');
    console.log('');

  } catch (error) {
    console.error('❌ Failed to set up database:', error);
    console.error('');
    console.error('Error details:', error.message);
    console.error('');
    console.error('Please ensure:');
    console.error('  1. You have run "npm install" to install dependencies');
    console.error('  2. The database/seed_complete.sql file exists');
    console.error('  3. You have write permissions to the database directory');
    console.error('');
    process.exit(1);
  }
}

setupDatabase();
